### ISSUE TITLE
Type error not caught when error contained within dataclass constructor?

### ISSUE THREAD

**Issue by jli** [microsoft/pylance-release#1087]
<h3> Environment data </h3>

-   Language Server version: 2021.3.2
-   OS and version: darwin x64
-   Python version (and distribution if applicable, e.g. Anaconda): 3.9.2
-   python.analysis.indexing: undefined
-   python.analysis.typeCheckingMode: basic

<h3> Expected behaviour </h3>

Expected type error would be flagged by Pylance.

<h3> Actual behaviour </h3>

Pylance doesn't raise any errors.

<h3> Logs </h3>

Python Language Server Log
```
[Info  - 8:09:47 PM] Pylance language server 2021.3.2 (pyright 85309906) starting
[Info  - 8:09:47 PM] Server root directory: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist
[Info  - 8:09:47 PM] No configuration file found.
[Info  - 8:09:47 PM] Setting pythonPath for service "pylance_errors": "/Users/johnli/bin/python"
Search paths found for configured python interpreter:
  /usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9
  /usr/local/Cellar/python@3.9/3.9.2_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/lib-dynload
  /usr/local/lib/python3.9/site-packages
[Error - 8:09:47 PM] stubPath /Users/johnli/src/pylance_errors/typings is not a valid directory.
[Info  - 8:09:47 PM] Assuming Python version 3.9
[Info  - 8:09:47 PM] Assuming Python platform Darwin
[Info  - 8:09:48 PM] Searching for source files
[Info  - 8:09:48 PM] Found 2 source files
[Info  - 8:09:48 PM] Background analysis(1) root directory: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist
[Info  - 8:09:48 PM] Background analysis(1) started
Background analysis message: setConfigOptions
Background analysis message: setTrackedFiles
[FG] parsing: /Users/johnli/src/pylance_errors/pylance_false_negative_dataclass.py (24ms)
[FG] parsing: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 3ms] (104ms)
[FG] binding: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist/typeshed-fallback/stdlib/builtins.pyi (43ms)
[FG] binding: /Users/johnli/src/pylance_errors/pylance_false_negative_dataclass.py (1ms)
Background analysis message: markAllFilesDirty
Background analysis message: ensurePartialStubPackages
Background analysis message: setFileOpened
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: analyze
[BG(1)] analyzing: /Users/johnli/src/pylance_errors/pylance_false_negative_dataclass.py ...
[BG(1)]   parsing: /Users/johnli/src/pylance_errors/pylance_false_negative_dataclass.py (25ms)
[BG(1)]   parsing: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 3ms] (99ms)
[BG(1)]   binding: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist/typeshed-fallback/stdlib/builtins.pyi (45ms)
[BG(1)]   binding: /Users/johnli/src/pylance_errors/pylance_false_negative_dataclass.py (0ms)
[BG(1)]   checking: /Users/johnli/src/pylance_errors/pylance_false_negative_dataclass.py ...
[BG(1)]     parsing: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist/typeshed-fallback/stdlib/dataclasses.pyi [fs read 1ms] (8ms)
[BG(1)]     binding: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist/typeshed-fallback/stdlib/dataclasses.pyi (1ms)
[BG(1)]     parsing: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi [fs read 1ms] (11ms)
[BG(1)]     binding: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi (2ms)
[BG(1)]     parsing: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist/typeshed-fallback/stdlib/typing.pyi [fs read 1ms] (31ms)
[BG(1)]     binding: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist/typeshed-fallback/stdlib/typing.pyi (12ms)
[BG(1)]     parsing: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist/typeshed-fallback/stubs/typing-extensions/typing_extensions.pyi [fs read 2ms] (5ms)
[BG(1)]     binding: /Users/johnli/.vscode/extensions/ms-python.vscode-pylance-2021.3.2/dist/typeshed-fallback/stubs/typing-extensions/typing_extensions.pyi (1ms)
[BG(1)]   checking: /Users/johnli/src/pylance_errors/pylance_false_negative_dataclass.py (116ms)
[BG(1)] analyzing: /Users/johnli/src/pylance_errors/pylance_false_negative_dataclass.py (288ms)
Background analysis message: resumeAnalysis
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
```

<h3> Code Snippet / Additional information </h3>

The 2 versions of "build_thing" are equivalent. Pylance finds the error with the 1st version (`build_thing__true_positive`), but doesn't find the error with the 2nd version (`build_thing__false_negative`).

```python
from dataclasses import dataclass

@dataclass
class Thing:
    str_field: str

def lookup_str(id: int, int_to_str: dict[int, str]) -> str:
    return int_to_str[id]

def build_thing__true_positive(int_to_int: dict[int, int]) -> Thing:
    # pylance correctly finds an error, underlining int_to_int:
    # > Argument of type "dict[int, int]" cannot be assigned to parameter
    # > "int_to_str" of type "dict[int, str]" in function "lookup_str"
    value = lookup_str(1, int_to_int)
    return Thing(value)

def build_thing__false_negative(int_to_int: dict[int, int]) -> Thing:
    # pylance doesn't find any errors in this version, however.
    return Thing(lookup_str(1, int_to_int))
```

mypy correctly finds both errors:

```shell
$ mypy pylance_false_negative_dataclass.py
pylance_false_negative_dataclass.py:14: error: Argument 2 to "lookup_str" has incompatible type "Dict[int, int]"; expected "Dict[int, str]"
pylance_false_negative_dataclass.py:19: error: Argument 2 to "lookup_str" has incompatible type "Dict[int, int]"; expected "Dict[int, str]"
Found 2 errors in 1 file (checked 1 source file)
```

**Comment by erictraut** [comment]
Thanks for the excellent bug report. Your repro case made it easy to find and diagnose the problem.

This will be fixed in the next release of pylance.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.3.4, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202134-31-march-2021
